### PR TITLE
Support 'spack external find' for rocm  packages(library only packages)

### PR DESCRIPTION
**Pull request by srekolam**
This is a follow up to the PR [#28005]
I added the support for the rocm packages.

### CODE REVIEW

**Review by bvanessen — DISMISSED** [review]

**Comment by alalazo** [comment]
@srekolam Can you rebase?

**Comment by alalazo** [comment]
@spackbot maintainers

**Comment by alalazo** [comment]
@cgmb Can you review this one?

File: var/spack/repos/builtin/packages/comgr/package.py [hunk 1]
```diff
@@ -16,6 +17,7 @@ class Comgr(CMakePackage):
     url      = "https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['libamd_comgr.so']
```
Comment (haampie) [inline var/spack/repos/builtin/packages/comgr/package.py]: @alalazo do we have something generic here? `.so/.dylib/.dll`?

Comment (bvanessen) [reply var/spack/repos/builtin/packages/comgr/package.py]: @haampie @alalazo @srekolam I think that the library names should not contain a suffix. During the search, spack should add the appropriate suffix based on the underlying OS.

**Comment by bvanessen** [comment]
@alalazo @haampie @srekolam So I think that there is an issue with the current infrastructure for finding libraries.  On a clean repo I can
```
[foo] (2)  %> spack external find --scope env:foo --all
==> The following specs have been detected on this system and added to 
... lots of packages but not miopen-hip
[foo] (2)  %> spack external find --scope env:foo miopen-hip
==> The following specs have been detected on this system and added to foo/spack.yaml
miopen-hip@5.1.0
[foo] (2) %> 
```

But I don't think that this is a problem for the PR to solve.

File: var/spack/repos/builtin/packages/comgr/package.py (hunk 1 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/comgr/package.py]: @haampie @alalazo @srekolam Okay, so I have verified that we can just put in the library without the suffix.  Spack will add it during the search.  However, the `determine_version` function is specific to finding based on the suffix available, which I think is fine.

Comment (bvanessen) [inline var/spack/repos/builtin/packages/comgr/package.py]:
```suggestion
    libraries = ['libamd_comgr']
```

File: var/spack/repos/builtin/packages/hip/package.py [hunk 2]
```diff
@@ -20,6 +21,8 @@ class Hip(CMakePackage):
     url      = "https://github.com/ROCm-Developer-Tools/HIP/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['libamdhip64.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/hip/package.py]:
```suggestion
    libraries = ['libamdhip64']
```

File: var/spack/repos/builtin/packages/hipsolver/package.py [hunk 3]
```diff
@@ -19,6 +21,7 @@ class Hipsolver(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/hipSOLVER/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam']
+    libraries = ['libhipsolver.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/hipsolver/package.py]:
```suggestion
    libraries = ['libhipsolver']
```

File: var/spack/repos/builtin/packages/hipsparse/package.py [hunk 4]
```diff
@@ -15,6 +17,7 @@ class Hipsparse(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/hipSPARSE/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['libhipsparse.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/hipsparse/package.py]:
```suggestion
    libraries = ['libhipsparse']
```

File: var/spack/repos/builtin/packages/hsa-rocr-dev/package.py [hunk 5]
```diff
@@ -20,6 +20,7 @@ class HsaRocrDev(CMakePackage):
     url      = "https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['libhsa-runtime64.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/hsa-rocr-dev/package.py]:
```suggestion
    libraries = ['libhsa-runtime64']
```

File: var/spack/repos/builtin/packages/migraphx/package.py [hunk 6]
```diff
@@ -14,6 +16,7 @@ class Migraphx(CMakePackage):
     url = "https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libmigraphx.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/migraphx/package.py]:
```suggestion
    libraries = ['libmigraphx']
```

File: var/spack/repos/builtin/packages/miopen-hip/package.py [hunk 7]
```diff
@@ -16,6 +17,7 @@ class MiopenHip(CMakePackage):
     url = "https://github.com/ROCmSoftwarePlatform/MIOpen/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libMIOpen.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/miopen-hip/package.py]:
```suggestion
    libraries = ['libMIOpen']
```

File: var/spack/repos/builtin/packages/miopen-opencl/package.py [hunk 8]
```diff
@@ -16,6 +17,7 @@ class MiopenOpencl(CMakePackage):
     url = "https://github.com/ROCmSoftwarePlatform/MIOpen/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libMIOpen.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/miopen-opencl/package.py]:
```suggestion
    libraries = ['libMIOpen']
```

File: var/spack/repos/builtin/packages/miopen-tensile/package.py [hunk 9]
```diff
@@ -16,6 +17,7 @@ class MiopenTensile(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/MIOpentensile/archive/rocm-5.0.0.tar.gz"
 
     maintainers = ['srekolam']
+    libraries = ['libMIOpenTensile.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/miopen-tensile/package.py]:
```suggestion
    libraries = ['libMIOpenTensile']
```

File: var/spack/repos/builtin/packages/miopengemm/package.py [hunk 10]
```diff
@@ -15,6 +17,7 @@ class Miopengemm(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/MIOpenGEMM/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libmiopengemm.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/miopengemm/package.py]:
```suggestion
    libraries = ['libmiopengemm']
```

File: var/spack/repos/builtin/packages/rdc/package.py [hunk 11]
```diff
@@ -14,6 +14,7 @@ class Rdc(CMakePackage):
     url      = "https://github.com/RadeonOpenCompute/rdc/archive/rocm-4.3.0.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librdc.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rdc/package.py]:
```suggestion
    libraries = ['librdc']
```

File: var/spack/repos/builtin/packages/rocalution/package.py [hunk 12]
```diff
@@ -21,6 +21,7 @@ class Rocalution(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocALUTION/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocalution_hip.so.0']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocalution/package.py]:
```suggestion
    libraries = ['librocalution_hip']
```

File: var/spack/repos/builtin/packages/rocblas/package.py [hunk 13]
```diff
@@ -15,6 +16,7 @@ class Rocblas(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocBLAS/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['librocblas.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocblas/package.py]:
```suggestion
    libraries = ['librocblas']
```

File: var/spack/repos/builtin/packages/rocfft/package.py [hunk 14]
```diff
@@ -15,6 +16,7 @@ class Rocfft(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocfft/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['librocfft.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocfft/package.py]:
```suggestion
    libraries = ['librocfft']
```

File: var/spack/repos/builtin/packages/rocm-dbgapi/package.py [hunk 15]
```diff
@@ -18,6 +19,7 @@ class RocmDbgapi(CMakePackage):
     url      = "https://github.com/ROCm-Developer-Tools/ROCdbgapi/archive/rocm-5.0.0.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocm-dbgapi.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocm-dbgapi/package.py]:
```suggestion
    libraries = ['librocm-dbgapi']
```

File: var/spack/repos/builtin/packages/rocm-debug-agent/package.py [hunk 16]
```diff
@@ -15,6 +16,7 @@ class RocmDebugAgent(CMakePackage):
     url      = "https://github.com/ROCm-Developer-Tools/rocr_debug_agent/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocm-debug-agent.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocm-debug-agent/package.py]:
```suggestion
    libraries = ['librocm-debug-agent']
```

File: var/spack/repos/builtin/packages/rocm-opencl/package.py [hunk 17]
```diff
@@ -13,6 +15,7 @@ class RocmOpencl(CMakePackage):
     git      = "https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime.git"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libOpenCL.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocm-opencl/package.py]:
```suggestion
    libraries = ['libOpenCL']
```

File: var/spack/repos/builtin/packages/rocm-smi-lib/package.py [hunk 18]
```diff
@@ -19,6 +20,7 @@ class RocmSmiLib(CMakePackage):
     url      = "https://github.com/RadeonOpenCompute/rocm_smi_lib/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocm_smi64.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocm-smi-lib/package.py]:
```suggestion
    libraries = ['librocm_smi64']
```

File: var/spack/repos/builtin/packages/rocprofiler-dev/package.py [hunk 19]
```diff
@@ -15,6 +16,7 @@ class RocprofilerDev(CMakePackage):
     url      = "https://github.com/ROCm-Developer-Tools/rocprofiler/archive/refs/tags/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocprofiler64.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocprofiler-dev/package.py]:
```suggestion
    libraries = ['librocprofiler64']
```

File: var/spack/repos/builtin/packages/rocrand/package.py [hunk 20]
```diff
@@ -18,6 +19,7 @@ class Rocrand(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocRAND/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocrand.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocrand/package.py]:
```suggestion
    libraries = ['librocrand']
```

File: var/spack/repos/builtin/packages/rocsolver/package.py [hunk 21]
```diff
@@ -17,6 +18,7 @@ class Rocsolver(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocSOLVER/archive/rocm-5.0.2.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
+    libraries = ['librocsolver.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocsolver/package.py]:
```suggestion
    libraries = ['librocsolver']
```

File: var/spack/repos/builtin/packages/rocsparse/package.py [hunk 22]
```diff
@@ -20,6 +21,7 @@ class Rocsparse(CMakePackage):
     url      = "https://github.com/ROCmSoftwarePlatform/rocSPARSE/archive/rocm-5.0.0.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['librocsparse.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/rocsparse/package.py]:
```suggestion
    libraries = ['librocsparse']
```

File: var/spack/repos/builtin/packages/roctracer-dev/package.py [hunk 23]
```diff
@@ -17,6 +18,7 @@ class RoctracerDev(CMakePackage):
     url      = "https://github.com/ROCm-Developer-Tools/roctracer/archive/rocm-5.1.3.tar.gz"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala']
+    libraries = ['libroctracer64.so']
```
Comment (bvanessen) [inline var/spack/repos/builtin/packages/roctracer-dev/package.py]:
```suggestion
    libraries = ['libroctracer64']
```

**Review by bvanessen — DISMISSED** [review]
Overall, looks good to me.  I did modify all of the libraries to remove the OS specific suffix.  I have tested on those packages that are installed on my system and it works fine.

**Comment by srekolam** [comment]
@bvanessen  thanks for the inputs. I made the changes to rccl and hipblas packages too as they had mentioned .so in libraries. if you can review that would be nice.

**Review by bvanessen — APPROVED** [review]

**Comment by srekolam** [comment]
@haampie or @tldahlgren , can you let me know your review comments if any ?

**Review by cgmb — COMMENTED** [review]
I'm not very familiar with this feature, but this looks pretty handy. In the future, you could probably detect which `amdgpu_target` the libraries are built for by looking for the embedded code object metadata.

**Review by alalazo — APPROVED** [review]

**Pull request merged**